Fix insertion of statements to be executed along return edge in inlining

tmiasko opened the pull request:
Inlining creates additional statements to be executed along the return
edge: an assignment to the destination, storage end for temporaries.

Previously those statements where inserted directly into a call target,
but this is incorrect when the target has other predecessors.

Avoid the issue by creating a new dedicated block for those statements.
When the block happens to be redundant it will be removed by CFG
simplification that follows inlining.

Fixes #117355

tmiasko committed:
Use let else to reduce indentation

cjgillot commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit d2ed699039ae59a9d8911adb63e7499b91f153e0 with merge af26da3cf9059da2b7aa0ac5641421a2d641b9c0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d2ed699039ae59a9d8911adb63e7499b91f153e0","merge_sha":"af26da3cf9059da2b7aa0ac5641421a2d641b9c0"} -->

cjgillot commented on compiler/rustc_mir_transform/src/inline.rs:
```diff
@@ -541,142 +540,158 @@ impl<'tcx> Inliner<'tcx> {
+        };
+
+        let return_block = if let Some(block) = target {
+            // Prepare a new block for code that should execute when call returns. We don't use
```
Should we attempt to avoid creating this new block if target has only a single predecessor? I started perf to verify whether there is anything to gain.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6857883574/job/18649421441)
Build commit: af26da3cf9059da2b7aa0ac5641421a2d641b9c0 (`af26da3cf9059da2b7aa0ac5641421a2d641b9c0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6857883574/job/18649421441"},"merge_sha":"af26da3cf9059da2b7aa0ac5641421a2d641b9c0"} -->

rust-timer commented:
Queued af26da3cf9059da2b7aa0ac5641421a2d641b9c0 with parent ba7c7a301984967c8c13adb580ef9b86ba706a83, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ba7c7a301984967c8c13adb580ef9b86ba706a83&end=af26da3cf9059da2b7aa0ac5641421a2d641b9c0).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (af26da3cf9059da2b7aa0ac5641421a2d641b9c0): [comparison URL](https://perf.rust-lang.org/compare.html?start=ba7c7a301984967c8c13adb580ef9b86ba706a83&end=af26da3cf9059da2b7aa0ac5641421a2d641b9c0&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=ba7c7a301984967c8c13adb580ef9b86ba706a83&end=af26da3cf9059da2b7aa0ac5641421a2d641b9c0&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ba7c7a301984967c8c13adb580ef9b86ba706a83&end=af26da3cf9059da2b7aa0ac5641421a2d641b9c0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8% | [0.8%, 0.8%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.8% | [0.8%, 0.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ba7c7a301984967c8c13adb580ef9b86ba706a83&end=af26da3cf9059da2b7aa0ac5641421a2d641b9c0&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.5% | [-0.5%, -0.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ba7c7a301984967c8c13adb580ef9b86ba706a83&end=af26da3cf9059da2b7aa0ac5641421a2d641b9c0&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 674.708s -> 673.584s (-0.17%)
**Artifact size**: 311.11 MiB -> 311.12 MiB (0.00%)

tmiasko committed:
Fix insertion of statements to be executed along return edge in inlining

Inlining creates additional statements to be executed along the return
edge: an assignment to the destination, storage end for temporaries.

Previously those statements where inserted directly into a call target,
but this is incorrect when the target has other predecessors.

Avoid the issue by creating a new dedicated block for those statements.
When the block happens to be redundant it will be removed by CFG
simplification that follows inlining.

Fixes #117355

cjgillot commented:
Apparently no perf to gain.
@bors r+

bors commented:
:pushpin: Commit d2ed699039ae59a9d8911adb63e7499b91f153e0 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot d2ed699039ae59a9d8911adb63e7499b91f153e0 -->
<!-- homu: {"type":"Approved","sha":"d2ed699039ae59a9d8911adb63e7499b91f153e0","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d2ed699039ae59a9d8911adb63e7499b91f153e0 with merge 1aa6312b2551c4e4ca60f9ba4c5b01824668df7b...
<!-- homu: {"type":"BuildStarted","head_sha":"d2ed699039ae59a9d8911adb63e7499b91f153e0","merge_sha":"1aa6312b2551c4e4ca60f9ba4c5b01824668df7b"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6922827985/job/18830002712) [(plain)](https://github.com/rust-lang-ci/rust/commit/1aa6312b2551c4e4ca60f9ba4c5b01824668df7b/checks/18830002712/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error: compilation failed!
status: exit status: 1
Build completed unsuccessfully in 0:11:17
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/mir-opt/inline/indirect_destination.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "-O" "-Copt-level=1" "-Zdump-mir=all" "-Zvalidate-mir" "-Zdump-mir-exclude-pass-number" "-Zmir-opt-level=0" "-Zmir-enable-passes=+Inline" "-Zdump-mir-dir=/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/inline/indirect_destination" "--emit" "mir" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/inline/indirect_destination" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/inline/indirect_destination/auxiliary" "--edition=2021"
--- stderr -------------------------------
error[E0061]: this function takes 3 arguments but 2 arguments were supplied
##[error]  --> /checkout/tests/mir-opt/inline/indirect_destination.rs:28:13
   |
   |
28 |             Call(*a = g(), bb1)
   |
note: function defined here
  --> /rustc/FAKE_PREFIX/library/core/src/intrinsics/mir.rs:318:24
help: provide the argument
help: provide the argument
   |
28 |             Call(*a = g(), bb1, /* unwind_action */)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0061`.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6922827985/job/18830137812)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6922827985/job/18830137812","builder_name":"checks-actions"} -->

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 1acecda95b7d83f0e838085a31c45a96c0aac0c3 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 1acecda95b7d83f0e838085a31c45a96c0aac0c3 -->
<!-- homu: {"type":"Approved","sha":"1acecda95b7d83f0e838085a31c45a96c0aac0c3","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 1acecda95b7d83f0e838085a31c45a96c0aac0c3 with merge db9e6558fdad8ae6957e8a815cc9b902f20eb29a...
<!-- homu: {"type":"BuildStarted","head_sha":"1acecda95b7d83f0e838085a31c45a96c0aac0c3","merge_sha":"db9e6558fdad8ae6957e8a815cc9b902f20eb29a"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6924526464/job/18833886143) [(plain)](https://github.com/rust-lang-ci/rust/commit/db9e6558fdad8ae6957e8a815cc9b902f20eb29a/checks/18833886143/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [mir-opt] tests/mir-opt/inline/exponential_runtime.rs stdout ----
69 +     }
70 + 
71 +     bb2: {
+ +         _4 = <() as F>::call() -> [return: bb1, unwind unreachable];
+ +     }
+ +     bb3: {
72 +         StorageDead(_7);
73 +         StorageDead(_6);
74 +         StorageDead(_5);
74 +         StorageDead(_5);

- +         _3 = <() as F>::call() -> [return: bb3, unwind unreachable];
+ +         _3 = <() as F>::call() -> [return: bb2, unwind unreachable];
77 + 
- +     bb3: {
- +     bb3: {
- +         _4 = <() as F>::call() -> [return: bb1, unwind unreachable];
+ +     bb4: {
+ +         _7 = <() as E>::call() -> [return: bb3, unwind unreachable];
81 + 
- +     bb4: {
+ +     bb5: {
83 +         StorageDead(_10);
83 +         StorageDead(_10);
84 +         StorageDead(_9);
85 +         StorageDead(_8);

- +         _6 = <() as E>::call() -> [return: bb5, unwind unreachable];
+ +         _6 = <() as E>::call() -> [return: bb4, unwind unreachable];
88 + 
- +     bb5: {
- +     bb5: {
- +         _7 = <() as E>::call() -> [return: bb2, unwind unreachable];
+ +     bb6: {
+ +         _10 = <() as D>::call() -> [return: bb5, unwind unreachable];
92 + 
- +     bb6: {
+ +     bb7: {
94 +         StorageDead(_13);
94 +         StorageDead(_13);
95 +         StorageDead(_12);
96 +         StorageDead(_11);

- +         _9 = <() as D>::call() -> [return: bb7, unwind unreachable];
+ +         _9 = <() as D>::call() -> [return: bb6, unwind unreachable];
99 + 
- +     bb7: {
- +     bb7: {
- +         _10 = <() as D>::call() -> [return: bb4, unwind unreachable];
+ +     bb8: {
+ +         _13 = <() as C>::call() -> [return: bb7, unwind unreachable];
103 + 
- +     bb8: {
+ +     bb9: {
105 +         StorageDead(_16);
105 +         StorageDead(_16);
106 +         StorageDead(_15);
107 +         StorageDead(_14);

- +         _12 = <() as C>::call() -> [return: bb9, unwind unreachable];
+ +         _12 = <() as C>::call() -> [return: bb8, unwind unreachable];
110 + 
- +     bb9: {
- +     bb9: {
- +         _13 = <() as C>::call() -> [return: bb6, unwind unreachable];
+ +     bb10: {
+ +         _16 = <() as B>::call() -> [return: bb9, unwind unreachable];
114 + 
- +     bb10: {
+ +     bb11: {
116 +         StorageDead(_19);
116 +         StorageDead(_19);
117 +         StorageDead(_18);
118 +         StorageDead(_17);

- +         _15 = <() as B>::call() -> [return: bb11, unwind unreachable];
+ +         _15 = <() as B>::call() -> [return: bb10, unwind unreachable];
121 + 
- +     bb11: {
- +     bb11: {
- +         _16 = <() as B>::call() -> [return: bb8, unwind unreachable];
- +     }
126 +     bb12: {
126 +     bb12: {
127 +         _18 = <() as A>::call() -> [return: bb13, unwind unreachable];

129 + 
130 +     bb13: {
130 +     bb13: {
- +         _19 = <() as A>::call() -> [return: bb10, unwind unreachable];
+ +         _19 = <() as A>::call() -> [return: bb11, unwind unreachable];
133   }
134   

thread '[mir-opt] tests/mir-opt/inline/exponential_runtime.rs' panicked at src/tools/compiletest/src/runtest.rs:4063:21:
thread '[mir-opt] tests/mir-opt/inline/exponential_runtime.rs' panicked at src/tools/compiletest/src/runtest.rs:4063:21:
Actual MIR output differs from expected MIR output /checkout/tests/mir-opt/inline/exponential_runtime.main.Inline.panic-abort.diff

---- [mir-opt] tests/mir-opt/inline/inline_coroutine.rs stdout ----
50       bb1: {
51 -         _3 = &mut _4;
51 -         _3 = &mut _4;
52 -         _2 = Pin::<&mut {coroutine@$DIR/inline_coroutine.rs:19:5: 19:8}>::new(move _3) -> [return: bb2, unwind unreachable];
- +         StorageDead(_7);
- +         StorageDead(_6);
- +         StorageDead(_5);
- +         StorageDead(_2);
- +         drop(_4) -> [return: bb2, unwind unreachable];
+ +         StorageDead(_4);
+ +         _0 = const ();
+ +         StorageDead(_1);
+ +         return;
59   
60       bb2: {

61 -         StorageDead(_3);
61 -         StorageDead(_3);
62 -         _1 = <{coroutine@$DIR/inline_coroutine.rs:19:5: 19:8} as Coroutine<bool>>::resume(move _2, const false) -> [return: bb3, unwind unreachable];
- +         StorageDead(_4);
- +         _0 = const ();
- +         StorageDead(_1);
- +         return;
+ +         StorageDead(_7);
+ +         StorageDead(_6);
+ +         StorageDead(_5);
+ +         StorageDead(_2);
+ +         drop(_4) -> [return: bb1, unwind unreachable];
68   
69       bb3: {

90 +     bb6: {
90 +     bb6: {
91 +         _1 = CoroutineState::<i32, bool>::Yielded(move _8);
92 +         discriminant((*_6)) = 3;
- +         goto -> bb1;
+ +         goto -> bb2;
95 + 
96 +     bb7: {

102 +         StorageDead(_8);
102 +         StorageDead(_8);
103 +         _1 = CoroutineState::<i32, bool>::Complete(_5);
104 +         discriminant((*_6)) = 1;
- +         goto -> bb1;
+ +         goto -> bb2;
107 + 
108 +     bb9: {

thread '[mir-opt] tests/mir-opt/inline/inline_coroutine.rs' panicked at src/tools/compiletest/src/runtest.rs:4063:21:
thread '[mir-opt] tests/mir-opt/inline/inline_coroutine.rs' panicked at src/tools/compiletest/src/runtest.rs:4063:21:
Actual MIR output differs from expected MIR output /checkout/tests/mir-opt/inline/inline_coroutine.main.Inline.panic-abort.diff

---- [mir-opt] tests/mir-opt/jump_threading.rs stdout ----
60       }
61   
62       bb1: {
-           StorageDead(_12);
-           StorageDead(_11);
-           StorageDead(_10);
-           StorageDead(_4);
-           _5 = discriminant(_3);
- -         switchInt(move _5) -> [0: bb2, 1: bb4, otherwise: bb3];
- +         goto -> bb2;
-   
-       bb2: {
73           StorageLive(_9);
73           StorageLive(_9);
74           _9 = ((_3 as Continue).0: i32);
75           _2 = _9;

77           _0 = Result::<i32, i32>::Ok(move _2);
78           StorageDead(_2);
79           StorageDead(_3);
-           goto -> bb5;
+           goto -> bb4;
82   
-       bb3: {
+       bb2: {
84           unreachable;
84           unreachable;
85       }
86   

-       bb4: {
+       bb3: {
88           StorageLive(_6);
89           _6 = ((_3 as Break).0: std::result::Result<std::convert::Infallible, i32>);
90           StorageLive(_8);
100           StorageDead(_6);
101           StorageDead(_2);
102           StorageDead(_3);
-           goto -> bb5;
---
107           return;
108       }
109   

+       bb5: {
+           StorageDead(_12);
+           StorageDead(_11);
+           StorageDead(_10);
+           StorageDead(_4);
+           _5 = discriminant(_3);
+ -         switchInt(move _5) -> [0: bb1, 1: bb3, otherwise: bb2];
+ +         goto -> bb1;
+   
110       bb6: {
110       bb6: {
111           _12 = move ((_4 as Err).0: i32);
112           StorageLive(_13);

113           _13 = Result::<Infallible, i32>::Err(move _12);
114           _3 = ControlFlow::<Result<Infallible, i32>, i32>::Break(move _13);
115           StorageDead(_13);
- -         goto -> bb1;
+ -         goto -> bb5;
117 +         goto -> bb9;
119   

124       bb8: {
124       bb8: {
125           _11 = move ((_4 as Ok).0: i32);
126           _3 = ControlFlow::<Result<Infallible, i32>, i32>::Continue(move _11);
-           goto -> bb1;
+           goto -> bb5;
129 + 
130 +     bb9: {

133 +         StorageDead(_10);
133 +         StorageDead(_10);
134 +         StorageDead(_4);
135 +         _5 = discriminant(_3);
- +         goto -> bb4;
+ +         goto -> bb3;
138   }
139   

thread '[mir-opt] tests/mir-opt/jump_threading.rs' panicked at src/tools/compiletest/src/runtest.rs:4063:21:
thread '[mir-opt] tests/mir-opt/jump_threading.rs' panicked at src/tools/compiletest/src/runtest.rs:4063:21:
Actual MIR output differs from expected MIR output /checkout/tests/mir-opt/jump_threading.identity.JumpThreading.panic-abort.diff

failures:
    [mir-opt] tests/mir-opt/inline/exponential_runtime.rs
    [mir-opt] tests/mir-opt/inline/inline_coroutine.rs
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6924526464/job/18834247293)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6924526464/job/18834247293","builder_name":"checks-actions"} -->

bors commented:
:pushpin: Commit 1be1c2ebcfd0ab10cbf001cc5d2471686a98aded has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 1be1c2ebcfd0ab10cbf001cc5d2471686a98aded -->
<!-- homu: {"type":"Approved","sha":"1be1c2ebcfd0ab10cbf001cc5d2471686a98aded","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 1be1c2ebcfd0ab10cbf001cc5d2471686a98aded with merge 79e961fa728e543e9b4f8b44f6bc2550ea867d4d...
<!-- homu: {"type":"BuildStarted","head_sha":"1be1c2ebcfd0ab10cbf001cc5d2471686a98aded","merge_sha":"79e961fa728e543e9b4f8b44f6bc2550ea867d4d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6928021545/job/18846963489)
Approved by: cjgillot
Pushing 79e961fa728e543e9b4f8b44f6bc2550ea867d4d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6928021545/job/18846963489"},"merge_sha":"79e961fa728e543e9b4f8b44f6bc2550ea867d4d"} -->

rust-timer commented:
Finished benchmarking commit (79e961fa728e543e9b4f8b44f6bc2550ea867d4d): [comparison URL](https://perf.rust-lang.org/compare.html?start=19079cf804330a0d3d767a19a935c5eebdfeba0f&end=79e961fa728e543e9b4f8b44f6bc2550ea867d4d&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=19079cf804330a0d3d767a19a935c5eebdfeba0f&end=79e961fa728e543e9b4f8b44f6bc2550ea867d4d&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=19079cf804330a0d3d767a19a935c5eebdfeba0f&end=79e961fa728e543e9b4f8b44f6bc2550ea867d4d&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 4.6% | [4.6%, 4.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=19079cf804330a0d3d767a19a935c5eebdfeba0f&end=79e961fa728e543e9b4f8b44f6bc2550ea867d4d&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.1% | [-4.6%, -2.1%] | 4     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=19079cf804330a0d3d767a19a935c5eebdfeba0f&end=79e961fa728e543e9b4f8b44f6bc2550ea867d4d&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.2%, -0.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

**Bootstrap**: 677.224s -> 674.774s (-0.36%)
**Artifact size**: 313.73 MiB -> 313.77 MiB (0.01%)
